picking parent in order of CollectionView, Table, ScrollView

yaphatak opened the pull request.

facebook-github-bot-7 commented:
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

facebook-github-bot-1 commented:
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

marekcirkos commented on WebDriverAgentLib/Categories/XCUIElement+FBScrolling.m:
```diff
@@ -80,9 +80,9 @@ - (BOOL)fb_scrollToVisibleWithNormalizedScrollDistance:(CGFloat)normalizedScroll
+  XCElementSnapshot *scrollView = [self.lastSnapshot fb_parentMatchingType:XCUIElementTypeCollectionView];
   scrollView = scrollView ?: [self.lastSnapshot fb_parentMatchingType:XCUIElementTypeTable];
-  scrollView = scrollView ?: [self.lastSnapshot fb_parentMatchingType:XCUIElementTypeCollectionView];
+  scrollView = scrollView ?: [self.lastSnapshot fb_parentMatchingType:XCUIElementTypeScrollView];
```
I think any chosen order will fail in some cases. Probably best would be to find parent that is either one of those three types.

yaphatak replied:
Yes you are right. I read your [comment](https://github.com/facebook/WebDriverAgent/issues/228#issuecomment-237214008) . Let me try to make this change. Because the current one only works if parent element is scrollview.

facebook-github-bot-9 commented:
@yaphatak updated the pull request.

marekcirkos commented on WebDriverAgentLib/Categories/XCElementSnapshot+Helpers.h:
```diff
@@ -38,6 +38,15 @@ NS_ASSUME_NONNULL_BEGIN
+ @return parent element matching one of given types
+ */
+
+- (nullable XCElementSnapshot *)fb_parentMatchingOneOfTypes:(NSArray<NSNumber *>*)types;
```
nit: `NSArray<NSNumber *> *`

marekcirkos commented on WebDriverAgentLib/Categories/XCElementSnapshot+Helpers.m:
```diff
@@ -24,6 +24,7 @@
 static NSString *const kXMLIndexPathKey = @"private_indexPath";
 
 inline static BOOL valuesAreEqual(id value1, id value2);
+inline static BOOL isSnapshotTypeAmongstGivenTypes(XCElementSnapshot* snapshot, NSArray<NSNumber *>* types);
```
nit:`(XCElementSnapshot *snapshot, NSArray<NSNumber *> *types`

marekcirkos commented on WebDriverAgentLib/Categories/XCElementSnapshot+Helpers.m:
```diff
@@ -96,6 +97,15 @@ - (XCElementSnapshot *)fb_parentMatchingType:(XCUIElementType)type
   return snapshot;
 }
 
+- (XCElementSnapshot *)fb_parentMatchingOneOfTypes:(NSArray<NSNumber *>*)types
```
nit: `(NSArray<NSNumber *> *)`

marekcirkos commented on WebDriverAgentLib/Categories/XCElementSnapshot+Helpers.m:
```diff
@@ -96,6 +97,15 @@ - (XCElementSnapshot *)fb_parentMatchingType:(XCUIElementType)type
+{
+ XCElementSnapshot *snapshot = self.parent;
+ while (snapshot && !isSnapshotTypeAmongstGivenTypes(snapshot, types)) {
+     snapshot = snapshot.parent;
```
nit: indentation should be two spaces

marekcirkos commented on WebDriverAgentLib/Categories/XCUIElement+FBScrolling.m:
```diff
@@ -80,9 +80,11 @@ - (BOOL)fb_scrollToVisibleWithNormalizedScrollDistance:(CGFloat)normalizedScroll
-  scrollView = scrollView ?: [self.lastSnapshot fb_parentMatchingType:XCUIElementTypeTable];
-  scrollView = scrollView ?: [self.lastSnapshot fb_parentMatchingType:XCUIElementTypeCollectionView];
+  NSArray *possibleParents = [NSArray arrayWithObjects:[NSNumber numberWithInteger:XCUIElementTypeScrollView],
+                                [NSNumber numberWithInteger:XCUIElementTypeCollectionView],
```
Could you use new objc notation for numbers and arrays like 
```
@[
  @(XCUIElementTypeScrollView),
  @(XCUIElementTypeCollectionView),
  @(XCUIElementTypeTable),
]`

marekcirkos commented:
In general looks good to me, however please fix indentation to two spaces and add some integration tests ;)

facebook-github-bot-7 commented:
@yaphatak updated the pull request.

yaphatak commented:
@marekcirkos : I fixed indentation and added objc annotation instead of Number. I see there are already couple integration tests for scrolling. You still want to add something specific?

marekcirkos commented:
I wouldn't test scrolling, but I definitely would test `fb_parentMatchingOneOfTypes` with some cases to make sure it works and definitely test how it works with `XCUIElementTypeAny. I am pretty sure it doesn't :)

facebook-github-bot-7 commented:
@yaphatak updated the pull request.

facebook-github-bot-8 commented:
@yaphatak updated the pull request.

yaphatak commented:
@marekcirkos : should i squash this so that you can merge?

marekcirkos commented:
There is no need to squash. Our bot does it anyway.

marekcirkos commented on WebDriverAgentLib/Categories/XCElementSnapshot+Helpers.h:
```diff
@@ -38,6 +38,15 @@ NS_ASSUME_NONNULL_BEGIN
+ @param types possible parent types
+ @return parent element matching one of given types
+ */
+
```
Kill empty line

marekcirkos commented on WebDriverAgentLib/Categories/XCElementSnapshot+Helpers.m:
```diff
@@ -118,3 +128,13 @@ inline static BOOL valuesAreEqual(id value1, id value2)
+
+inline static BOOL isSnapshotTypeAmongstGivenTypes(XCElementSnapshot* snapshot, NSArray<NSNumber *> *types)
+{
+  for (NSUInteger i = 0; i < types.count; i++) {
```
You might also want to check if `type[i]` is `XCUIElementTypeAny`.

marekcirkos commented on WebDriverAgentTests/IntegrationTests/XCElementSnapshotHelperTests.m:
```diff
@@ -64,4 +64,18 @@ - (void)testParentMatchingType
 }
 
+- (void)testParentMatchingOneOfTypes
+{
```
Please split those two cases and add one more test for request of non existing type.

facebook-github-bot-6 commented:
@yaphatak updated the pull request.

marekcirkos commented:
This is nice. Thank you!

marekcirkos commented:
@facebook-github-bot shipit

facebook-github-bot-8 commented:
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3696275) to review internal test results.
